internal/envoy/bootstrap.go

```diff
@@ -141,6 +173,81 @@ func upstreamFileTLSContext(cafile, certfile, keyfile string) *envoy_api_v2_auth
 	return context
 }
 
+func upstreamSdsTLSContext(certificateSdsFile, validationSdsFile string) *envoy_api_v2_auth.UpstreamTlsContext {
+	context := &envoy_api_v2_auth.UpstreamTlsContext{
+		CommonTlsContext: &envoy_api_v2_auth.CommonTlsContext{
+			TlsCertificateSdsSecretConfigs: []*envoy_api_v2_auth.SdsSecretConfig{{
+				SdsConfig: &envoy_api_v2_core.ConfigSource{
+					ConfigSourceSpecifier: &envoy_api_v2_core.ConfigSource_Path{
+						Path: certificateSdsFile,
+					},
+				},
+			}},
+			ValidationContextType: &envoy_api_v2_auth.CommonTlsContext_ValidationContextSdsSecretConfig{
+				ValidationContextSdsSecretConfig: &envoy_api_v2_auth.SdsSecretConfig{
+					SdsConfig: &envoy_api_v2_core.ConfigSource{
+						ConfigSourceSpecifier: &envoy_api_v2_core.ConfigSource_Path{
+							Path: validationSdsFile,
+						},
+					},
+				},
+			},
+		},
+	}
+	return context
+}
+
+func discoveryResponse(secret *envoy_api_v2_auth.Secret) *api.DiscoveryResponse {
+	response := &api.DiscoveryResponse{
+		Resources: []*any.Any{toAny(secret)},
+	}
+	return response
+}
+
+// tlsCertificateSdsSecretConfig creates DiscoveryResponse with file based SDS resource
+// including paths to TLS certificates and key
+func tlsCertificateSdsSecretConfig(certfile, keyfile string) *api.DiscoveryResponse {
```

Comment:
If this function, `validationContextSdsSecretConfig` and the internal `boostrapConfig` one all take a `*BootstrapConfig` argument, then you can have a consistent function signature for all of them, which will make the caller pretty clean (it can just loop through a slice of function pointers).

Reply:
I aligned the arguments, but the return types of the three functions need to be specific types since they are used by calling code, so I could not align the return types to `proto.Message`.

There is also surrounding logic in `Bootstrap()` where these function are called from, related to how/when to call, how to use return values, so I could not figure out a way how to call them in a loop as function pointers.

Writing out the resulting `proto.Message` buffers as config files in `WriteBootstrap()` is now a loop though.

Reply:
I like the refactoring that you have done here. I think that if we take it a bit further, it cleans up even more

https://gist.github.com/jpeach/3a0e331181357408b544c1185037f749

LMK what you think?

Reply:
Thanks, did this too and adjusted the test case accordingly.